### ISSUE TITLE
3.0 Queries: Base query class (meta, dates, caching)

### ISSUE THREAD

**Issue by JJJ** [easydigitaldownloads/easy-digital-downloads#6331]
Tracking issue: #4576

Pull request: #6333 

# TL;DR

Transition Payments/Discounts/Logs/Notes/Orders/Order-Items post-types to custom component database tables, most paired with meta. Architect a traditional WordPress query interface that's simple and DRY for all of them to maintain developer happiness.

# Summary

One technical challenge for EDD3 is the introduction of approximately 12 new database tables (for a total of 14.) The reason for this move is to ultimately provide a more scalable application without a _requirement_ for third-party horizontal scaling techniques (multiple servers; WordPress drop-in plugins; cache/db/lb/web/etc...)

By scaling the databases horizontally by default, EDD can better control the flow of data internally without compromising performance. Put another way, by not storing everything as a custom post type & post-meta, EDD can avoid hurting the performance of other post-type related data, while also improving its own efficiency.

It's a win-win for everyone, but with a hidden caveat...

We learned first-hand (via BuddyPress and Multisite) that the above approach works well-enough at providing a stable foundation for scaling a complex WordPress application, but we also learned having more than a half-dozen different interfaces for interacting with new & unique components in WordPress is a hugely negative experience for traditional WordPress developers.

Each new component/meta relationship is a new hurdle to jump, and too many hurdles leaves developers exhausted and unhappy.

I'm proposing (and working on) a mostly-magic base query-class for all custom database tables to extend, that will automatically make every new table queryable using the same basic parameters, provide object caching for those individual items, without compromising either application performance or developer joy.

Kinda like an [ORM](https://en.wikipedia.org/wiki/Object-relational_mapping), but without the need to relearn the WordPress way of querying for things.

## How does it work?

1. Component classes provide an array of column definitions and other component attributes
1. Column definitions help determine the shape of the component item
1. That shape helps define the queryable columns
1. Queries are cached by an md5 hash of parameters used
1. Items are cached by their ID

# Components

Here is the master list of components, alphabetically. Each will get checked when complete:

- [x] Customers
- [x] Discounts
- [x] Logs (General)
- [x] Logs (API Requests, No Meta)
- [x] Logs (Downloads, No Meta)
- [x] Notes
- [x] Orders
- [x] Order Items (Linked to Orders)

# Orders (Previously Payments)

Table/schema completed (see #6277)

### Query vars
- [x] `id`
- [x] `id__in`
- [x] `id__not_on`
- [x] `number`
- [x] `number__in`
- [x] `number__not_in`
- [x] `status`
- [x] `status__in`
- [x] `status__not_in`
- [x] `user_id`
- [x] `user__in`
- [x] `user__not_in`
- [x] `customer_id`
- [x] `customer__in`
- [x] `customer__not_in`
- [x] `email`
- [x] `email__in`
- [x] `email__not_in`
- [x] `gateway`
- [x] `gateway__in`
- [x] `gateway__not_in`
- [x] `payment_key`
- [x] `payment_key__in`
- [x] `payment_key__not_in`
- [x] `date_created_query`
- [x] `date_completed_query`
- [x] `meta_query`

**Comment by pippinsplugins** [comment]
Add `gateway`, `gateway_in`, and `gateway__not_in` please.

**Comment by JJJ** [comment]
> Add `gateway`, `gateway_in`, and `gateway__not_in` please.

Added.

**Comment by cklosowski** [comment]
@jjj looks like `gateway` came out as `gateway_key` in the list above, is that intended?

**Comment by JJJ** [comment]
>  looks like gateway came out as gateway_key in the list above, is that intended?

Copy/pasta. Fixed.

**Comment by JJJ** [comment]
Crud; I included a file that was part of a different stash, so I'm gonna clean that up quick.

**Comment by JJJ** [comment]
OK, here's where this is at now:

* `wp_edd_orders` has full support for querying single/__in/__not_in most columns
* `date_created` and `date_completed` have `WP_Date_Query` support
* `meta_query` support is in via `WP_Meta_Query`
* The base `EDD_DB_Query` class is the base class that other query classes can extend. It automatically sets up what type of queries are possible based on database column definitions. It's a lot like `EDD_DB` but on steroids to be more DRY
* There's a lot to like, and a lot to be critical of, so reviews from @sunnyratilal @cklosowski @pippinsplugins are encouraged. Very much WIP still, emphasis on W.

**Comment by cklosowski** [comment]
@jjj getting this error when hitting 'checkout'
```
( ! ) Fatal error: Uncaught Error: Call to a member function get_column() on null in /app/public/wp-content/plugins/easy-digital-downloads/includes/payments/class-edd-payment.php on line 2352
```

**Comment by cklosowski** [comment]
@jjj seems to be related to the `customer` component. It's returning `null` instead of the `customer` component when calling `EDD()->customers`. Still digging in to figure out why.

**Comment by sunnyratilal** [comment]
⚠️ As part of the new customers query class (and after discussion with @JJJ), the `exists()` method that is part of `EDD_DB_Customers` will no longer exist and will be a breaking change in 3.0.  This is something that we will need to note in the developer notes.

This will fail gracefully eventually when we introduce a shim class that handles all the "breaking" changes in 3DD.

**Comment by sunnyratilal** [comment]
Closing out as this has been completed. Any bugs found during alpha/beta testing should be opened as new issues.
